[결제] 토스결제검증 로직 작성 #46

zmfpdl64 opened the pull request:
## 개요
branch: #46 
사전에 생성한 객체와 비교하여 토스에게 승인하는 로직입니다.

```mermaid
sequenceDiagram

Client ->> Toss: 사전결제

Toss ->> Client: redirect 전송

Client ->> Payment: 결제 확인객체 전달(성공)

Payment ->> Client: 결제 검증(실패)

Payment ->> User: 빈페이 추가 요청

User ->> Payment: 예외 반환(실패)

Payment ->> Client: 실패 반환

User ->> Payment: 성공 반환

Payment ->> Toss: 결제 승인

Toss ->> Payment: 결제 승인 실패

Payment ->> Client: 결제 실패

Toss ->> Payment: 결제 승인 반환

Payment ->> Client: 성공반환
```
## PR 유형
어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [x] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [x] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [x] 테스트 추가, 테스트 리팩토링
- [x] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.  Commit message convention 참고  (Ctrl + 클릭하세요.) 
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

> ### Description
- [x] 설정파일 작성 ✅ 2024-04-15
- [x] 수정해야할 코드 Refactor ✅ 2024-04-15
- [x] 검증 로직 작성 ✅ 2024-04-15
    - [x] 컨트롤러 ✅ 2024-04-15
    - [x] 서비스 ✅ 2024-04-15
    - [x] 레포지토리 ✅ 2024-04-15
    - [x] 예외 ✅ 2024-04-15
    - [x] Entity ✅ 2024-04-15
    - [x] Dto ✅ 2024-04-15


 - [x] 컨트롤러 테스트 ✅ 2024-04-15
- [x] 서비스 테스트 ✅ 2024-04-15
	- [x] 정상 테스트 통과 ✅ 2024-04-15
	- [x] 빈페이 존재  x ✅ 2024-04-15
	- [x] 빈페이 검증 값 불일치 ✅ 2024-04-15
	- [x] 토스 검증 승인 예외 ✅ 2024-04-15
- [x] dto 테스트 ✅ 2024-04-15
	- [x] 성공 ✅ 2024-04-15
	- [x] 예외 케이스 ✅ 2024-04-15
- [x] Entity 테스트 ✅ 2024-04-15
	- [x] 진행중 ✅ 2024-04-15
	- [x] 검증 ✅ 2024-04-15
	- [x] 완료 ✅ 2024-04-15
	- [x] 실패 ✅ 2024-04-15







> ### Core Code

```java
	@Transactional
	public BeanPayDto.Response.TossPayment validTossCharge(BeanPayDto.Request.TossPayment request) {
		log.info("request : {} {} {}", request.paymentKey(), request.orderId(), request.amount());

		BeanPay findBeanPay = findBeanPayById(request.orderId());

		try {
			processInProgress(findBeanPay);
			validateBeanPay(findBeanPay, request);
		} catch (CustomException e) {
			handleException(findBeanPay, e);
			throw e;
		}
		try {
			//TODO: 유저 beanPay 추가 로직 작성
			return processApproval(findBeanPay, request);
		} catch (CustomException e) {
			handleException(findBeanPay, e);
			//TODO: 유저 beanPay 롤백 로직 작성
			throw e;
		}
	}
```

zmfpdl64 committed:
Chore: valid 의존성과 spring cloud 버전 수정, toss SecretKey 적용

1. Valid 의존성 추가
2. Spring Cloud 버저닝
3. Toss Secret Key 적용

zmfpdl64 committed:
Refactor: 빈페이 필드 추가

결제 로직을 수행하면 부수적인 필드가 필요하다고 생각됨
1. 결제 방식
2. 취소 이유
3. 승인 시점
4. ID -> UUID 전환
UUID로 전환한 이유는 토스로 충전 ID를 전달할 때 6자리 이상의 값을 전달해야하기 때문

zmfpdl64 committed:
Feat: 토스 비밀키 생성

application.yml의 값을 먼저 읽어드리고 TossKey 컴포넌트를 생성해야 하기 때문에 ConfigProperties 설정 클래스를 만듬

zmfpdl64 committed:
Refactor: api versioning

api에 버전까지 확인하기 위해 url path를 변경

자원을 생성하는 로직이라 get이 아닌 post가 적절하여 수정

zmfpdl64 committed:
Feat: 토스 Feign 클래스

토스 검증로직을 위해 Feign 작성

zmfpdl64 committed:
Refactor: beanPay ID -> UUID로 변경

토스페이에 충전ID를 전달해야 하는데 이때 6자리 이상의 ID를 전달해야 한다. 그래서 기존의 ID방식이 아닌 UUID방식으로 변경했다.

zmfpdl64 committed:
Feat: String 변환 LocalDateTime

토스의 날짜 형식에 맞춰 LocalDateTime으로 변환해주는 Util 클래스

zmfpdl64 committed:
Feat: 충전로직 검증 기능 구현

Service: Transactional을 클래스단에 위치시켰다.

zmfpdl64 committed:
Test: 토스결제승인 검증로직 테스트 작성

hjm12223 commented on payment-api/src/main/java/org/ecommerce/paymentapi/entity/BeanPay.java:
```diff
@@ -29,8 +35,10 @@
 public class BeanPay {
 
 	@Id
-	@GeneratedValue(strategy = GenerationType.IDENTITY)
-	private Long id;
+	@GeneratedValue(generator = "uuid2")
+	@GenericGenerator(name = "uuid2", strategy = "uuid2")
+	@Column(columnDefinition = "BINARY(16)")
+	private UUID id;
 
```
혹시 UUID를 키값으로 하신 이유가 토스 페이먼츠 API와 연관이 있어서 인가용?

hjm12223 commented on payment-api/src/test/java/org/ecommerce/paymentapi/service/BeanPayServiceTest.java:
```diff
@@ -26,23 +35,140 @@ class BeanPayServiceTest {
+
+			final BeanPayDto.Request.TossPayment request = new BeanPayDto.Request.TossPayment(paymentType, paymentKey,
+				orderId, amount);
+			final BeanPay entity = new BeanPay(orderId, null, userId, amount, null, null, BeanPayStatus.DEPOSIT,
+				ProcessStatus.PENDING, LocalDateTime.now(), null);
```
혹시 BeanPay.ofCreate()를 안사용하신 이유가 있을까요?

hjm12223 commented:
수고많으셨습니다 우진님! 테스트 코드를 엄청 깔끔하게 잘 작성하신거 같습니다 저도 참고하도록 하겠습니다 ㅎㅎ

zmfpdl64 committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feature/payment/#46

zmfpdl64 committed:
Refactor: 주문 검증 로직 수정

fail 메소드에서 fail or cancle 이유를 에러코드를 받아 처리해 유연성을 증가 시킴

try/catch 문을 적용하여 롤백 후 fail메소드를 수행하여 상태값을 변경시킴

변경된 로직에 맞춰 테스트 코드 수정

zmfpdl64 committed:
Refactor: validBeanPay 매개변수 변경

확장성을 위하여 DTO로 받지 않고 매개변수를 값으로 변경

Joowon-Seo commented on payment-api/src/main/java/org/ecommerce/paymentapi/service/BeanPayService.java:
```diff
@@ -11,21 +16,64 @@
+		//빈페이 찾기
+		final BeanPay findBeanPay = beanPayRepository.findById(request.orderId())
+			.orElseThrow(() -> new CustomException(BeanPayErrorCode.NOT_EXIST));
+		try{
+			// beanPay 진행중 상태 변경
+			findBeanPay.inProgress();
+
+			// 빈페이 값 검증하기
+			if (!findBeanPay.validBeanPay(request.orderId(), request.amount())) {
+				throw new CustomException(BeanPayErrorCode.VERIFICATION_FAIL);
+			}
+		}catch (CustomException e) {
+			findBeanPay.fail(e.getErrorCode());
+			throw e;
+		}
+
+		try{
+			//TODO: 유저에게 beanPay 전달
+
+			//결제 승인
+			final ResponseEntity<BeanPayDto.Response.TossPayment> response = tossServiceClient.approvePayment(tossKey.getAuthorizationKey(),  request);
 
+			// 승인 예외
+			if(!response.getStatusCode().is2xxSuccessful()){
+				log.error("토스 결제 승인 실패");
+				throw new CustomException(BeanPayErrorCode.TOSS_RESPONSE_FAIL);
+			}
+
+			//충전객체 상태변경
+			findBeanPay.complete(response.getBody());
+			log.info("토스 결제 승인 서비스 로직 종료");
+			return response.getBody();
+
+		}catch(CustomException e) {
+
+			findBeanPay.fail(e.getErrorCode());
+			//TODO: 유저에게 beanPay 롤백
+			throw e;
+		}
+	}
```
안녕하세요 우진님! 주석을 잘 달아주셔서 이해하기 편했습니다 ㅎㅎ

`validTossCharge()` 메소드에 다양한 비즈니스 로직이 있어서 SPR 원칙에 위배될 수 있고, Transaction의 범위가 넓은 것 같다는 생각이 듭니다.

beanpay를 찾거나 검증하는 부분은 Transaction의 범위에서 빠질 수 있을 것 같습니다.
try-catch로 분리해주신 것 처럼 별도의 메소드로 분리하여 관리하면 어떨까용?
상태 변경을 DB에 로깅하는 작업도 수월할 것 같습니다!

Joowon-Seo commented on payment-api/src/main/java/org/ecommerce/paymentapi/utils/TossKey.java:
```diff
@@ -0,0 +1,34 @@
+public class TossKey {
+
+	private final ConfigProperties properties;
+	@Getter
+	private final String authorizationKey;
+
```
`@Getter`는 어디에서 사용되나용?

Joowon-Seo reviewed (approved):
결제 어렵네요... 😢
고생많으셨습니다!! 👍 👍

zmfpdl64 committed:
Refactor: 토스결제검증 로직 메소드 단위로 분리

zmfpdl64 committed:
Refactor: 비밀키 접근자 제한

생성자가 실행될 때 한번 실행되면 되기 떄문에 private으로 접근제어자 변경했다!

zmfpdl64 committed:
Docs: TODO 주석 작성

zmfpdl64 committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feature/payment/#46

no-oneho commented on .idea/gradle.xml:
```diff
@@ -14,6 +14,7 @@
             <option value="$PROJECT_DIR$/common" />
             <option value="$PROJECT_DIR$/order-api" />
             <option value="$PROJECT_DIR$/payment-api" />
+            <option value="$PROJECT_DIR$/product" />
```
저 포함 다들 실수하면서 그래들 xml 파일이 자꾸 올라옵니다 ㅜㅜ git 추적을 해제하고 이그노어에 올리던지 하는게 좋을것같습니다 한번 방법을 다같이 찾아봅시다!

no-oneho commented on payment-api/src/main/java/org/ecommerce/paymentapi/config/ConfigProperties.java:
```diff
@@ -0,0 +1,18 @@
+@Setter
+@Validated
+@ConfigurationProperties("toss")
+public class ConfigProperties {
+	private String secretKey;
+}
```
이 부분은 record로 작성되어도 괜찮아보입니다. 관련 링크 첨부해드립니다!
https://stackoverflow.com/questions/66696828/how-to-use-configurationproperties-with-records

no-oneho commented on payment-api/src/main/java/org/ecommerce/paymentapi/controller/BeanPayController.java:
```diff
@@ -5,26 +5,34 @@
+	@GetMapping("/success")
+	public Response<BeanPayDto.Response.TossPayment> validCharge(@Valid final BeanPayDto.Request.TossPayment request) {
+		final BeanPayDto.Response.TossPayment response = beanPayService.validTossCharge(request);
+		return new Response<>(200, response);
```
200 보단 HttpStatus.OK.value() 가 더 직관적일것같습니다! .value() 부분은 공통단 수정해서 HttpStatus.OK만 쓸수있게 추후 수정하면 될것같습니다

no-oneho commented on payment-api/src/main/java/org/ecommerce/paymentapi/entity/BeanPay.java:
```diff
@@ -54,6 +69,9 @@ public class BeanPay {
 	@Column(name = "create_datetime", updatable = false)
 	private LocalDateTime createDateTime;
 
+	@Column(name = "approve_datetime", updatable = false)
```
updateable = false 속성은 update 쿼리에 들어가지않는 컬럼으로 지정하는 속성인데
혹시 approve_datetime 이 그런 컬럼인가요? 컬럼명만 보면 승인시간같아서 update때 들어가지않나? 생각이 들어서 혹시나하고 코멘트 남깁니다!

no-oneho commented on payment-api/src/main/java/org/ecommerce/paymentapi/exception/BeanPayErrorCode.java:
```diff
@@ -0,0 +1,25 @@
+
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor
```
getter 어노테이션 추가하시면 아래 오버라이드 메서드들 지우셔도 될 것 같습니다!

no-oneho reviewed (approved):
자세하게 코드랑 간간히 주석도 달려 이해하기 편했습니다! 고생많으셨습니다!
